safer cleanup when failing to compile regexps

tonycoz opened the pull request:
Prior to this commit when producing a warning the regexp compiler would check if the warning category was marked as FATAL, and if it was it would add clean up to the save stack to release buffers used during compilation and to release the working REGEXP SV.
    
This causes two type of problems:
    
- if an error was already queued, Perl_ck_warner() returns even if the warning is fatal, this meant that the normal clean up code Perl_re_op_compile() would also run, resulting in a double free of the buffers.
    
- without fatal warnings, if a $SIG{__WARN__} handler died, the buffers and the working REGEXP SV would leak.
    
Avoid this by using SAVEDESTRUCTOR_X() to release the memory and optionally the SV at the end of scope.
    
Fixes #21661

tonycoz committed:
safer cleanup when failing to compile regexps

Prior to this commit when producing a warning the regexp compiler
would check if the warning category was marked as FATAL, and if it was
it would add clean up to the save stack to release buffers used during
compilation and to release the working REGEXP SV.

This causes two type of problems:

- if an error was already queued, Perl_ck_warner() returns even if
  the warning is fatal, this meant that the normal clean up code
  Perl_re_op_compile() would also run, resulting in a double free
  of the buffers.

- without fatal warnings, if a $SIG{__WARN__} handler died, the
  buffers and the working REGEXP SV would leak.

Avoid this by using SAVEDESTRUCTOR_X() to release the memory and
optionally the SV at the end of scope.

Fixes #21661

tonycoz committed:
add tests for #21661

mauke commented on regcomp.c:
```diff
@@ -1213,6 +1213,32 @@ S_set_regex_pv(pTHX_ RExC_state_t *pRExC_state, REGEXP *Rx)
+release_RExC_state(pTHX_ void *vstate) {
+    RExC_state_t *pRExC_state = (RExC_state_t *)vstate;
+
+    if (RExC_rx_sv)                                                     \
+        SvREFCNT_dec(RExC_rx_sv);                                         \
```
Why backslashes at the end of the line?

mauke replied:
I believe this whole block could be replaced by just `SvREFCNT_dec(RExC_rx_sv);` or even `SvREFCNT_dec_set_NULL(RExC_rx_sv);`.

mauke commented on regcomp.c:
```diff
@@ -1213,6 +1213,32 @@ S_set_regex_pv(pTHX_ RExC_state_t *pRExC_state, REGEXP *Rx)
+
+    if (RExC_rx_sv)                                                     \
+        SvREFCNT_dec(RExC_rx_sv);                                         \
+    if (RExC_open_parens) {
```
Aren't all of these `if` checks redundant? `free` should be a no-op on a NULL pointer.

mauke commented on regcomp_internal.h:
```diff
@@ -919,7 +918,6 @@ static const scan_data_t zero_scan_data = {
  * Calls SAVEDESTRUCTOR_X if needed, then Simple_vFAIL()
  */
 #define vFAIL(m) STMT_START {                           \
-    PREPARE_TO_DIE;                                     \
     Simple_vFAIL(m);                                    \
 } STMT_END
```
This could be simplified to:
```c
#define vFAIL(m) Simple_vFAIL(m)
```

Ditto (mutatis mutandis) for `vFAIL2`, `vFAIL3`, `vFAIL4`, `vFAIL2utf8f`, `vFAIL3utf8f` below. (Alternatively, the `vFAIL` wrappers could be removed entirely.)

mauke commented on regcomp_internal.h:
```diff
@@ -919,7 +918,6 @@ static const scan_data_t zero_scan_data = {
  * Calls SAVEDESTRUCTOR_X if needed, then Simple_vFAIL()
```
This comment is no longer accurate because the macro no longer calls `SAVEDESTRUCTOR_X`. Same for all the following `vFAIL` variants.

tonycoz committed:
regcomp_internal.h: simplify vFAIL*() macros

These can be simpler, and since we allow use of __VA_ARGS__
we can consolidate Simple_vFAIL[1-4]() into Simple_vFAILn()

tonycoz committed:
regcomp_internal.h: removed unused PREPARE_TO_DIE

We are now always prepared for death.

demerphq commented:
@tonycoz looks like this can be merged?

jkeenan commented:
@tonycoz, this p.r. has had merge conflicts for close to 5 months.  Do you intend to go forward with it?  If not, please consider closing it.  Thanks.

mauke commented:
I've rebased the PR on top of blead and removed the conflicts. Unless someone objects (or beats me to the punch), I intend on merging this within a day or two.
